### PR TITLE
[PVR] Cleanup texture db and stale cached images

### PR DESCRIPTION
**Pull request by ksooo**
Texture db and `Thumbnails` directory over time gets filled by PVR components with lots of stale db entries and cached files for channel logos, EPG tag thumbnails, recording thumbnails, recording icons, ...

This PR introduces a background job to cleanup the mess on a regular base (30 secs after PVR startup, then every 12 hours). 

Although for some user actions we could detect the need to clean the db immediately (e.g. user triggers to delete a recording from inside Kodi), we also need a background job to catch deletion/change of images done by the user in the backend while kodi is not running. The job covers everything, no need to pollute the code with additional logics to do a live cleanup. 

Changes were tested carefully for some weeks on macOS and Android.

@phunkyfish when you find some time for a review, I suggest to do it commit by commit, thaen you should be able to follow the code logics quite well.

### CODE REVIEW

File: xbmc/pvr/epg/Epg.cpp
Comment (neo1973) [inline xbmc/pvr/epg/Epg.cpp:561]: What's the use of `paths` and the following loop? Wouldn't assigning directly to `urlsToCheck` do the same thing?

Comment (ksooo) [reply xbmc/pvr/epg/Epg.cpp:561]: Good catch. Thanks. Must be a leftover from an earlier iteration of this PR.

File: xbmc/pvr/channels/PVRChannel.h
Comment (phunkyfish) [inline xbmc/pvr/channels/PVRChannel.h:40]: Default value in a constructor definition. This feels incorrect. It would be cleaner to have 2 constructors, one with iconPath and one without.

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (phunkyfish) [inline xbmc/pvr/epg/EpgDatabase.cpp:1013]: Is this a standard pattern for DB reads? If not, catch everything makes me worry we're not sure what could go wrong.

**Review by phunkyfish — CHANGES_REQUESTED** [review]
Just really the constructor

File: xbmc/pvr/epg/EpgDatabase.cpp
Comment (ksooo) [reply xbmc/pvr/epg/EpgDatabase.cpp:1013]: Yes, it's like this everywhere in PVR and EPG database. Introduced long before my Kodi times.

Comment (phunkyfish) [reply xbmc/pvr/epg/EpgDatabase.cpp:1013]: Ok, cool.

File: xbmc/pvr/channels/PVRChannel.h
Comment (ksooo) [reply xbmc/pvr/channels/PVRChannel.h:40]: Okay, will change that.

**Comment by neo1973** [comment]
This goes into bikeshedding territory so feel free to ignore this comment.

I noticed this pattern a couple of times:
```c++
class T
{
private:
  std::string m_str;
public:
  T(const std::string& str) : m_str(str) {}
};
```
In cases like this it's beneficial to take the constructors parameter as value instead of reference. Because with a reference, the compiler has to make a copy in every case but with the value move semantic can avoid the copy in some circumstances. So the constructor is more efficient this way:
```c++
T(str::string str) : m_str(std::move(str)) {}
```

**Comment by ksooo** [comment]
> T(str::string str) : m_str(std::move(str)) {}

 I know this pattern and tend not to use it.

**Comment by ksooo** [comment]
@phunkyfish I did the requested changes, please approve if you feel this is okay now.

**Comment by ksooo** [comment]
@neo1973 you seem to a clue about C++. I invite you to open PRs for the conceptional changes you propose here and there in different PRs as side notes.

**Review by phunkyfish — APPROVED** [review]

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
@phunkyfish once this PR got merged, you should extend your PVR Providers PR to support also texture cache cleanup. I saw that a Provider has an attribute "iconpath", thus the icon will be cached in the texture db once it gets used by any UI component.

**Comment by phunkyfish** [comment]
> @phunkyfish once this PR got merged, you should extend your PVR Providers PR to support also texture cache cleanup. I saw that a Provider has an attribute "iconpath", thus the icon will be cached in the texture db once it gets used by any UI component.

Great idea. I’ll have a look tomorrow.

**Comment by ksooo** [comment]
Well, jenkins is acting up, but errors are not related to this PR.

**Comment by phunkyfish** [comment]
> @phunkyfish once this PR got merged, you should extend your PVR Providers PR to support also texture cache cleanup. I saw that a Provider has an attribute "iconpath", thus the icon will be cached in the texture db once it gets used by any UI component.

What specifically should I integrate? I.e. what is the set of steps I should take?

**Comment by ksooo** [comment]
> Similar to what is done in the channels commit?

 yes.

**Comment by ksooo** [comment]
Ping me if you have something. I will be happy to review.

**Comment by phunkyfish** [comment]
> Ping me if you have something. I will be happy to review.

It possible the provider PR will get merged before I add support. We’ve been waiting a long while to get it merged.

**Comment by ksooo** [comment]
> It possible the provider PR will get merged before I add support. We’ve been waiting a long while to get it merged.

Sure. Go ahead.

**Pull request merged**
